inspektor-gadget/inspektor-gadget
# gadgets: Introduce bpfstats gadget

This PR introduces support for exposing statistics (metrics) about the ebpf programs and gadgets running on the system. 
Check the individual commits to get more details on the implementation.

## Testing 

1. build the gadget

```bash 
$ sudo ig image build gadgets/bpfstats -t bpfstats
```

2. start ig on daemon mode 

```bash 
$ sudo ig daemon --verify-image=false --enable-bpfstats
```

3. Run some gadgets

```bash 
$ sudo gadgetctl run --name mytraceopen trace_open:latest --detach
$ sudo gadgetctl run --name mytraceexec trace_exec:latest --detach
```

### Using the CLI 

4. Run the bpfstats gadget 

```bash 
$ sudo gadgetctl run bpfstats --sort -runtime
WARN[0000] local                | image signature verification is disabled due to using corresponding option
GADGETID    GADGETNAME           GADGETIMAGE          PROGID PROGNAME                     RUNTIME        RUNCOUNT       MAPMEMORY MAPCO…
a9b10cf8fd5 mytraceopen          trace_open:latest         0                               100497             258         1710240      6
9da0de911e9 mytraceexec          trace_exec:latest         0                                    0               0       183170336      7
```

By default, the gadget shows the stats aggregated by gadget. Using `--programs` makes it return per-program stats:

```bash
$ sudo gadgetctl run bpfstats --sort -runtime --programs
GADGETID    GADGETNAME           GADGETIMAGE          PROGID PROGNAME                     RUNTIME        RUNCOUNT       MAPMEMORY MAPCO…
                                                          51 __x64_sys_socke             55538314           10478         6467272      7
                                                          50 __x64_sys_socke             18184594           10474         1083840      4
                                                          53 __x64_sys_bind_             15679318            4407         6467272      7
                                                          47 __x64_sys_unlin             12154534            1278         6467272      7
                                                          55 __x64_sys_conne             11895276            3375         6467272      7
                                                          52 __x64_sys_bind_              5629158            4407         1083840      4
                                                          39 sched_process_e              4845671            2760         4391808      4
                                                          38 task_newtask                 4841680            2475         4391808      4
                                                          54 __x64_sys_conne              3883108            3341         1083840      4
                                                          46 __x64_sys_unlin              3576277            1278         1083840      4
                                                          41 __x64_sys_renam              3163412             286         6467272      7
                                                          37 sched_process_e              1597635             256         4391808      4
                                                          49 __x64_sys_unlin              1490552             225         6467272      7
a9b10cf8fd5 mytraceopen          trace_open:latest       643 ig_openat_e                  1185322            3154         1098240      3
                                                          59 __x64_sys_accep              1106122             242         6467272      
...
```

### Exporting to Prometheus

Enable the metrics on the daemon

```bash
$ sudo ig daemon --verify-image=false --enable-bpfstats --otel-metrics-listen
```

now run the gadget with the annotations to enable the metrics

```bash 
$ sudo gadgetctl run bpfstats --otel-metrics-name=bpfstats:bpfstats --annotate=bpfstats:metrics.collect=true --detach
INFO[0000] installed as "8423cd8e53339c8d4501ec7cdff436bc"
```

check the available metrics:

```bash 
$ curl http://localhost:2224/metrics -s  | grep bpfstats
mapCount{gadgetImage="trace_exec:latest",gadgetName="mytraceexec",otel_scope_name="bpfstats",otel_scope_version="",progName=""} 7
mapCount{gadgetImage="trace_open:latest",gadgetName="mytraceopen",otel_scope_name="bpfstats",otel_scope_version="",progName=""} 6
mapMemory{gadgetImage="trace_exec:latest",gadgetName="mytraceexec",otel_scope_name="bpfstats",otel_scope_version="",progName=""} 1.83170336e+08
mapMemory{gadgetImage="trace_open:latest",gadgetName="mytraceopen",otel_scope_name="bpfstats",otel_scope_version="",progName=""} 1.71024e+06
otel_scope_info{otel_scope_name="bpfstats",otel_scope_version=""} 1
runcount_total{gadgetImage="trace_exec:latest",gadgetName="mytraceexec",otel_scope_name="bpfstats",otel_scope_version="",progName=""} 636
runcount_total{gadgetImage="trace_open:latest",gadgetName="mytraceopen",otel_scope_name="bpfstats",otel_scope_version="",progName=""} 219513
runtime_total{gadgetImage="trace_exec:latest",gadgetName="mytraceexec",otel_scope_name="bpfstats",otel_scope_version="",progName=""} 445114
runtime_total{gadgetImage="trace_open:latest",gadgetName="mytraceopen",otel_scope_name="bpfstats",otel_scope_version="",progName=""} 7.3383934e+07
```

### Discussion

- Should it be really a gadget? in this case, almost all the work is done by the operator, the gadget is just a template enabling it. I was able to implement a [POC](https://github.com/inspektor-gadget/inspektor-gadget/tree/mauricio/bpfstats-without-gadget) to have this as a first-class citizen in IG:

```bash 
$ sudo gadgetctl stats
preRun
GADGETID    GADGETNAME           GADGETIMAGE          PROGID PROGNAME                     RUNTIME        RUNCOUNT       MAPMEMORY MAPCO…
GADGETID    GADGETNAME           GADGETIMAGE          PROGID PROGNAME                     RUNTIME        RUNCOUNT       MAPMEMORY MAPCO…
GADGETID    GADGETNAME           GADGETIMAGE          PROGID PROGNAME                     RUNTIME        RUNCOUNT       MAPMEMORY MAPCO…
a9b10cf8fd5 mytraceopen          trace_open:latest         0                              2194531            7884         1710240      6
9da0de911e9 mytraceexec          trace_exec:latest         0                                23245              28       183170336      7
```

### TODO


### Issues to open
- [ ] bpf_metadata (https://github.com/inspektor-gadget/inspektor-gadget/pull/3820#discussion_r1969477020)
- [ ] How to hide columns based on params (https://github.com/inspektor-gadget/inspektor-gadget/pull/3820#discussion_r1951757640)
- [ ] eBPF stats are never disabled 
- [ ] Gadget name and ID missing for attached gadgets (https://github.com/inspektor-gadget/inspektor-gadget/pull/3820#discussion_r2035553258)
- [ ] Provide human version of runtime and map size
- [ ] Test external programs too

Fixes #3759
